Add `cleanup` option

dignifiedquire opened the pull request:
As spawned processes are not cleaned up when the main process dies
this option adds the abilitiy to kill those when the spawning process
dies.

Closes #51

dignifiedquire commented:
Looks like this is not working on CI :( passed fine for me locally will investigate what the issue is.

sindresorhus commented on index.js:
```diff
@@ -33,7 +34,8 @@ function handleArgs(cmd, args, opts) {
 		encoding: 'utf8',
-		reject: true
+		reject: true,
+		cleanup: false
```
Any reason we wouldn't want this as `true` by default? I already assumed that's how the core one worked. We could clearly document it in the readme.

dignifiedquire replied:
I was just conscious of this being a breaking change and thought you might not want that, happy to default it to `true`.

sindresorhus replied:
We already have other breaking changes in master. So will be a breaking release regardless.

dignifiedquire replied:
Okay, will change then

sindresorhus commented on readme.md:
```diff
@@ -11,7 +11,7 @@
 - [Improved Windows support.](https://github.com/IndigoUnited/node-cross-spawn#why)
 - Higher max buffer. 10 MB instead of 200 KB.
 - [Executes locally installed binaries by name.](#preferlocal)
-
```
Don't remove the newline

sindresorhus commented on readme.md:
```diff
@@ -11,7 +11,7 @@
 - Higher max buffer. 10 MB instead of 200 KB.
 - [Executes locally installed binaries by name.](#preferlocal)
-
+- [Cleans up spawned processes when the parent process dies](#cleanup)
```
Should end in a `.`

sindresorhus commented on readme.md:
```diff
@@ -138,6 +138,12 @@ Default: `true`
+Type: `boolean`<br>
+Default: `false`
+
+Setting this to `true` keeps track of the spawned process and `kill`s it when the parent process exits.
```
> Keep track of the spawned process and `kill`s it when the parent process exits.

sindresorhus commented on test.js:
```diff
@@ -276,3 +277,24 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
+async function spawnAndKill(t, signal) {
+	const cp = m('sub-process');
+	let pid;
+
```
`cp.stdout.setEncoding('utf8')`

Then you don't have to `.toString()` it below.

sindresorhus commented on test.js:
```diff
@@ -276,3 +277,24 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
+		pid = parseInt(chunk.toString(), 10);
+		t.is(typeof pid, 'number');
+
+		setTimeout(function () {
```
use arrow function

sindresorhus commented on fixtures/sub-process:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env node
+'use strict';
+
```
remove this newline

dignifiedquire commented:
@sindresorhus squashed commits and travis seems to be happy now that I introduced a small delay for the pid check.

I've also added a test to make sure it behaves as expected when using `cleanup: false`

sindresorhus commented on test.js:
```diff
@@ -276,3 +277,54 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
+	await t.throws(cp);
+
+	// Give everybody some time to breath and kill things
+	const timeout = delay => new Promise(resolve => setTimeout(resolve, delay));
```
Just use https://github.com/sindresorhus/delay

dignifiedquire replied:
I knew there was a module somewhere ;)

sindresorhus commented on test.js:
```diff
@@ -276,3 +277,54 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
+	// Give everybody some time to breath and kill things
+	const timeout = delay => new Promise(resolve => setTimeout(resolve, delay));
+
+	await timeout(100);
```
100=>200 just to be safe

sindresorhus commented on test.js:
```diff
@@ -276,3 +277,54 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
+		setTimeout(() => {
+			process.kill(cp.pid, 'SIGINT');
+		}, 100);
+	});
```
This is almost a duplicate of the other test. Can you extract this into a function both could use?

dignifiedquire commented:
Squashed and green (at least before squash)

dignifiedquire commented:
@sindresorhus ping

sindresorhus commented:
Thank you! :)

Sorry about the delay. I've been quite busy.

Unfortunately we can't do a new release with this until https://github.com/sindresorhus/execa/issues/46 is resolved... Any help welcome.

dignifiedquire commented:
> Sorry about the delay. I've been quite busy.

No need to be sorry :) Thank you for merging
